[CPS] Support in Discover

mbondyra opened the pull request:
## Summary

Adds CPS functionality to Discover, enabling saved searches and Discover embeddables to react to changes in CPS project picker.
<img width="1058" height="591" alt="Screenshot 2025-12-03 at 15 01 09" src="https://github.com/user-attachments/assets/420c725e-9574-4b67-b6e4-13a1542cc8a8" />

### Discover app
1. When going to Discover, project routing is set to default spaces setting (for now hardcoded).
2. When moving to another app or reloading Discover page, the setting is reset.
3. Project routing is not persisted in URL.
4. The user has the ability to save the project routing setting to a Discover session. If it's saved, next time it will be loaded with the saved setting.
5. Project routing is persisted per discover session if the user choses to save it to saved object, not per tab (with the ability to override for ESQL tabs)

### Dashboard
1. By default, Discover session takes the CPS setting from the Dashboard (implemented [here](https://github.com/elastic/kibana/pull/243569))
2. If Discover session has its own project routing saved in, it takes precedence over the Dashboard setting and informs the user about scope overrides via custom badge (please do not pay much attention to the badge current design, it will be refined later on + it doesn't display yet for ESQL cases):
<img width="738" height="459" alt="Screenshot 2025-12-03 at 14 28 39" src="https://github.com/user-attachments/assets/c3139d1f-de6f-44f5-9330-37e10c25b25a" />


Add one more thing for Dashboard - When coming back from embeddable editor to the dashboard, if there is CPS scope saved to dashboard, the dashboard loads the saved CPS scope. Otherwise It remains on the previously active CPS scope from embeddable.

mbondyra committed:
[CPS] Discover + embeddable

mbondyra committed:
null/undefined cleanup

kibanamachine committed:
Changes from node scripts/jest_integration -u src/core/server/integration_tests/ci_checks

mbondyra committed:
null/undefined cleanup

mbondyra committed:
null undefined more cleanups

kibanamachine committed:
Changes from node scripts/eslint_all_files --no-cache --fix

nreese commented:
Any way to make the CPS override badge in the panel header smaller and  look more like CPS picker button from unified search?

mbondyra committed:
fixtures (?)

jughosta commented on src/platform/plugins/shared/discover/public/application/main/components/top_nav/save_discover_session/save_modal.tsx:
```diff
@@ -68,22 +81,15 @@ export const DiscoverSessionSaveModal: React.FC<DiscoverSessionSaveModalProps> =
           <FormattedMessage
             id="discover.topNav.saveModal.storeTimeWithSearchToggleLabel"
-            defaultMessage="Store time with Discover session"
+            defaultMessage="Include time range and refresh interval customizations"
```
It reads like 2 verbs with `and` in between: `include` and `refresh` which makes it confusing

jughosta replied:
```suggestion
            defaultMessage="Include time range customizations"
```

jughosta commented on src/platform/plugins/shared/saved_search/public/service/save_discover_session.ts:
```diff
@@ -23,7 +23,9 @@ export type SaveDiscoverSessionParams = Pick<
-  Partial<Pick<DiscoverSession, 'id'>>;
+  Partial<Pick<DiscoverSession, 'id'>> & {
+    projectRouting?: DiscoverSession['projectRouting'] | null;
+  };
```
Can we also "pick" `projectRouting` from `DiscoverSession` instead?

mbondyra replied:
I don't think so. `projectRouting` from the persisted saved object shape includes null apart from type DiscoverSession['projectRouting'] stores. Maybe I am missing something, but I don't think it's possible. Happy to accept contributions :)

davismcphee replied:
I'd prefer not to have saved object client `null`/`undefined` weirdness surfaced in application code if we can help it. As far as app code goes, I think it's cleaner to just deal with `ProjectRouting | undefined`.

But I also don't understand why we can't pick `projectRouting` from `DiscoverSession` here. We already coerce `undefined` to `null` when saving below, so why would we need to accept `null` in `SaveDiscoverSessionParams`?

jughosta commented on src/platform/plugins/shared/saved_search/server/saved_objects/schema.ts:
```diff
@@ -202,5 +202,31 @@ export const SCHEMA_SEARCH_MODEL_VERSION_9_SO_API_WORKAROUND = schema.object({
+// We need to extend from the base props rather than the full schema to avoid the same error as version 8
+const SCHEMA_SEARCH_MODEL_VERSION_10_EXTENDED = SCHEMA_SEARCH_MODEL_VERSION_8.extends({
+  projectRouting: schema.maybe(schema.nullable(schema.string())),
+  tabs: schema.arrayOf(SCHEMA_DISCOVER_SESSION_TAB_VERSION_10, { minSize: 1 }),
```
Do we have to mention `tabs`?

davismcphee replied:
We don't have a fix yet for the SO API issue so I think we still need to work around it, but it seems like we could simplify some. Would something like this work instead?
```ts
export const SCHEMA_SEARCH_MODEL_VERSION_10 = SCHEMA_SEARCH_MODEL_VERSION_8.extends({
  projectRouting: schema.maybe(schema.nullable(schema.string())),
});

const { tabs: tabsV10, ...restV10Props } = SCHEMA_SEARCH_MODEL_VERSION_10.getPropSchemas();

export const SCHEMA_SEARCH_MODEL_VERSION_10_SO_API_WORKAROUND = schema.object({
  ...restV10Props,
  tabs: schema.maybe(tabsV10),
});
```

I don't think it's necessary to change `DiscoverSessionTabAttributes` or `DiscoverSessionTab` either.

jughosta commented on src/platform/plugins/shared/discover/public/application/main/state_management/redux/actions/reset_discover_session.ts:
```diff
@@ -94,6 +94,20 @@ export const resetDiscoverSession = createInternalStateAsyncThunk(
+        // This preserves the picker value after "Save As" when projectRouting is not stored
+        const defaultRouting = services.cps.cpsManager.getDefaultProjectRouting();
+        dispatch(internalStateSlice.actions.setProjectRouting(defaultRouting));
+        services.cps.cpsManager.setProjectRouting(defaultRouting);
```
We have many places now there these two actions are called. Can we rather have one place/listener which calls them when `persistedDiscoverSession` changes in the redux store?

mbondyra replied:
If we drop it, we'll have to pass cpsService to more places (harder testing).  We'll have to create a new action for tab marking, and it is actually much more lines of changes. But up to you, please just let me know whichever you prefer.

davismcphee replied:
Hmm this one's on me, I originally said it was fine to put in the Redux store. I'd typically advise against relying on global services directly for state management since it's created challenges for tabs (and future features like split view). But it might be different in this case since it's shared across all tabs.

I think a single source of truth is ideal though. We could either rely on the global service directly, or it might even be easier / fewer code changes to do 2-way syncing with the service like we do with the app state: https://github.com/elastic/kibana/blob/a4679a0c43b65dd09e79d1c527298c4055c92de1/src/platform/plugins/shared/discover/public/application/main/state_management/discover_state.ts#L544-L552

That way it could live in the Redux store as far as Discover's concerned and be easily accessible, but also stay in sync with the global service.

mbondyra committed:
### Dashboard

When carrying some context from embeddable to the dashboard, if there is CPS scope saved to dashboard, the dashboard does not reset to its default state. It remains on the previously active CPS scope from embeddable (can be observed for Lens)

jughosta reviewed (commented):
Thanks for contributing to Discover code! Leaving some initial comments.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/369527)
* Commit: de70014c483844ad361294d101f4fdb667a6e5d0

### Failed CI Steps
* [Jest Tests #1](https://buildkite.com/elastic/kibana-pull-request/builds/369527#019aeef1-f8bd-4cb9-b0b9-02281663a69d)
* [Jest Tests #1](https://buildkite.com/elastic/kibana-pull-request/builds/369527#019aef04-8992-45fd-aabd-36021336b89f)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/369527#019aeef1-f8bd-4cb9-b0b9-02281663a69d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/369527/jobs/019aeef1-f8bd-4cb9-b0b9-02281663a69d/artifacts/019aef04-68a5-49f6-911a-800fccf446cd) Jest Tests #<span></span>1 / projectRouting Should detect projectRouting change from _alias:_origin to ALL when projectRoutingRestore is true
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/369527#019aef04-8992-45fd-aabd-36021336b89f) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/369527/jobs/019aef04-8992-45fd-aabd-36021336b89f/artifacts/019aef15-e129-41d7-b598-6c062dd033a4) Jest Tests #<span></span>1 / projectRouting Should detect projectRouting change from _alias:_origin to ALL when projectRoutingRestore is true
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/369527#019aeef1-f8bd-4cb9-b0b9-02281663a69d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/369527/jobs/019aeef1-f8bd-4cb9-b0b9-02281663a69d/artifacts/019aef04-68a6-4e92-8e36-b95a8c3e49f6) Jest Tests #<span></span>1 / projectRouting Should distinguish between ALL (saved with all projects) and undefined (not saved)
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/369527#019aef04-8992-45fd-aabd-36021336b89f) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/369527/jobs/019aef04-8992-45fd-aabd-36021336b89f/artifacts/019aef15-e129-49ba-a534-dabf6f9f8dcd) Jest Tests #<span></span>1 / projectRouting Should distinguish between ALL (saved with all projects) and undefined (not saved)

### History
* :green_heart: [Build #368732](https://buildkite.com/elastic/kibana-pull-request/builds/368732) succeeded 0cd869900037dbfd32693a9bf3d256f64357f30e
* :broken_heart: [Build #368672](https://buildkite.com/elastic/kibana-pull-request/builds/368672) failed 3a872d2d417ed0430e965b4a1fa38ae7f92a864d
* :broken_heart: [Build #368491](https://buildkite.com/elastic/kibana-pull-request/builds/368491) failed fee12440ab7599e511996afa2c9e393b0873ff69
* :broken_heart: [Build #368338](https://buildkite.com/elastic/kibana-pull-request/builds/368338) failed c87efbe0e0f6ce078cb550d2c995d0b1a32a412e
* :broken_heart: [Build #368206](https://buildkite.com/elastic/kibana-pull-request/builds/368206) failed a8ccd3aac28fd149d04c28233c139e84307cf56a

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369527","number":369527,"commit":"de70014c483844ad361294d101f4fdb667a6e5d0"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368732","number":368732,"commit":"0cd869900037dbfd32693a9bf3d256f64357f30e"},{"buildStatus":{"state":"running","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368672","number":368672,"commit":"3a872d2d417ed0430e965b4a1fa38ae7f92a864d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368491","number":368491,"commit":"fee12440ab7599e511996afa2c9e393b0873ff69"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368338","number":368338,"commit":"c87efbe0e0f6ce078cb550d2c995d0b1a32a412e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368206","number":368206,"commit":"a8ccd3aac28fd149d04c28233c139e84307cf56a"}],"number":369527}
buildkite-pr-comment-->

mbondyra committed:
cr

mbondyra committed:
test

mbondyra committed:
remove from state

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/369702)
* Commit: 09491897287058baf52970d6250751c2a8f96158

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/1e7ce545fca220cd45baa905255a330bb33c7e60) | [after](https://github.com/elastic/kibana/commit/09491897287058baf52970d6250751c2a8f96158) | diff |
| --- | --- | --- | --- |
| `discover` | 1862 | 1863 | +1 |
| `presentationPanel` | 159 | 160 | +1 |
| total |  |  | +2 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/1e7ce545fca220cd45baa905255a330bb33c7e60) | [after](https://github.com/elastic/kibana/commit/09491897287058baf52970d6250751c2a8f96158) | diff |
| --- | --- | --- | --- |
| `discover` | 157 | 158 | +1 |
| `savedSearch` | 102 | 103 | +1 |
| total |  |  | +2 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/1e7ce545fca220cd45baa905255a330bb33c7e60) | [after](https://github.com/elastic/kibana/commit/09491897287058baf52970d6250751c2a8f96158) | diff |
| --- | --- | --- | --- |
| `aiops` | 526.6KB | 526.8KB | +172.0B |
| `dashboard` | 702.3KB | 702.3KB | +23.0B |
| `dataVisualizer` | 597.5KB | 597.6KB | +170.0B |
| `discover` | 1.3MB | 1.3MB | +2.8KB |
| `embeddableAlertsTable` | 1001.0KB | 1001.2KB | +173.0B |
| `lens` | 1.6MB | 1.6MB | +178.0B |
| `maps` | 3.1MB | 3.1MB | +179.0B |
| `ml` | 5.6MB | 5.6MB | +344.0B |
| `presentationPanel` | 40.7KB | 43.0KB | +2.3KB |
| `savedSearch` | 6.9KB | 7.1KB | +216.0B |
| `slo` | 991.7KB | 991.9KB | +172.0B |
| `synthetics` | 1.0MB | 1.0MB | +172.0B |
| `visualizations` | 343.5KB | 343.7KB | +172.0B |
| total |  |  | +7.0KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/1e7ce545fca220cd45baa905255a330bb33c7e60) | [after](https://github.com/elastic/kibana/commit/09491897287058baf52970d6250751c2a8f96158) | diff |
| --- | --- | --- | --- |
| `presentationPanel` | 10.1KB | 10.3KB | +198.0B |
| `savedSearch` | 10.1KB | 10.3KB | +113.0B |
| total |  |  | +311.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/1e7ce545fca220cd45baa905255a330bb33c7e60) | [after](https://github.com/elastic/kibana/commit/09491897287058baf52970d6250751c2a8f96158) | diff |
| --- | --- | --- | --- |
| `discover` | 199 | 200 | +1 |
| `savedSearch` | 106 | 107 | +1 |
| total |  |  | +2 |


</details>


### History
* :broken_heart: [Build #369527](https://buildkite.com/elastic/kibana-pull-request/builds/369527) failed de70014c483844ad361294d101f4fdb667a6e5d0
* :green_heart: [Build #368732](https://buildkite.com/elastic/kibana-pull-request/builds/368732) succeeded 0cd869900037dbfd32693a9bf3d256f64357f30e
* :broken_heart: [Build #368672](https://buildkite.com/elastic/kibana-pull-request/builds/368672) failed 3a872d2d417ed0430e965b4a1fa38ae7f92a864d
* :broken_heart: [Build #368491](https://buildkite.com/elastic/kibana-pull-request/builds/368491) failed fee12440ab7599e511996afa2c9e393b0873ff69
* :broken_heart: [Build #368338](https://buildkite.com/elastic/kibana-pull-request/builds/368338) failed c87efbe0e0f6ce078cb550d2c995d0b1a32a412e

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369702","number":369702,"commit":"09491897287058baf52970d6250751c2a8f96158"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369527","number":369527,"commit":"de70014c483844ad361294d101f4fdb667a6e5d0"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368732","number":368732,"commit":"0cd869900037dbfd32693a9bf3d256f64357f30e"},{"buildStatus":{"state":"running","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368672","number":368672,"commit":"3a872d2d417ed0430e965b4a1fa38ae7f92a864d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368491","number":368491,"commit":"fee12440ab7599e511996afa2c9e393b0873ff69"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368338","number":368338,"commit":"c87efbe0e0f6ce078cb550d2c995d0b1a32a412e"}],"number":369702}
buildkite-pr-comment-->

davismcphee commented on packages/kbn-check-saved-objects-cli/src/migrations/__fixtures__/search/10.10.0.json:
.

davismcphee commented on src/platform/plugins/shared/discover/public/application/main/state_management/redux/selectors/unsaved_changes.ts:
```diff
@@ -61,6 +61,18 @@ export const selectHasUnsavedChanges = (
+      before: persistedProjectRouting,
+      after: state.projectRouting,
+    });
+  }
```
I might be misunderstanding, but we don't compare saved search attributes here, we use `DiscoverSessionTab`. I don't think we have a pattern for comparing top-level session attributes yet though, so might need to do it explicitly.

davismcphee commented on src/platform/plugins/shared/discover/public/application/main/data_fetching/fetch_all.ts:
```diff
@@ -97,6 +98,7 @@ export function fetchAll(
         services,
         sort: sort as SortOrder[],
         inputTimeRange: currentTab.dataRequestParams.timeRangeAbsolute,
+        projectRouting: internalState.getState().projectRouting,
```
It doesn't seem like we're syncing it to saved search in Discover atm. Not sure I'd invest much effort in trying to either since the plan is to fully remove it from Discover's state. It does feel a bit odd to pull only one value from Redux like this in the meantime, but might be fine.

davismcphee replied:
It doesn't seem like we're syncing it to saved search in Discover atm. Not sure I'd invest much effort in trying to either since the plan is to fully remove it from Discover's state. It does feel a bit odd to pull only one value from Redux like this in the meantime, but might be fine.

davismcphee commented on src/platform/plugins/shared/discover/public/application/main/data_fetching/fetch_all.ts:
Yeah that's a really good point... @mbondyra Do we have a plan for this in general for this project? Seems like we'd have similar issues in other apps too. Almost feels like we should centralize injecting the global project routing value to requests if not overridden, maybe in the data plugin or something. Not sure how feasible something like that actually is.

mbondyra replied:
Hi Davis,
Yeah, so we actually thought about this before, but at the time it didn’t seem like there were many cases. With the previous design (overriding layers in Lens + picker in Unified Search), it felt simpler to pass the projectRouting from cpsManager → embeddable → data plugin.
However, now that it seems we’re only supporting the navigation picker (no overrides needed), so based on this comment and seeing the complexity, let's have the data plugin inject the `projectRouting` value, and let consumers only subscribe to global picker changes to refetch. This should help with all the search calls. PR here: https://github.com/elastic/kibana/pull/245746

As for the rest, we’re writing it down (thanks @jughosta for being so thorough!), but these won’t be supported in the tech preview — we have to have ES routes working first. Anyway, I'll move this PR to draft because parts of it will be rewritten on top of https://github.com/elastic/kibana/pull/245746

davismcphee reviewed (commented):
Just responding to some review comments.

nreese commented on src/platform/plugins/shared/dashboard/public/dashboard_api/project_routing_manager.ts:
```diff
@@ -19,28 +19,28 @@ export const COMPARE_DEBOUNCE = 100;
   initialState: DashboardState,
-  projectRoutingRestore$: PublishingSubject<boolean>
+  projectRoutingRestore$: PublishingSubject<boolean>,
+  hasIncomingEmbeddables?: boolean
```
Looks like `hasIncomingEmbeddables` is not used? Could this be removed?
